### PR TITLE
Leverage ActivityListener.AutoGenerateRootContextTraceId

### PR DESCRIPTION
**Pull request by cijothomas**
Fixes portions of #941 #953 

## Changes
Leverage ActivityListener.AutoGenerateRootContextTraceId so that sampling callbacks can get the actual TraceId of the to-be-created Activity and make decision based on TraceId. 
Before this PR, Samplers generated a random TraceId, and made sampling decision. But this generated TraceID was not actually used when the Activity is created, leading to broken traces, if samplers made decision based on the TraceId.

Added Tests. There are TODOs which require follow up.

Please provide a brief description of the changes here. Update the
`CHANGELOG.md` for non-trivial changes.

For significant contributions please make sure you have completed the following items:

* [ ] Design discussion issue #
* [ ] Changes in public API reviewed

### CODE REVIEW

File: src/OpenTelemetry/CHANGELOG.md [hunk 1]
```diff
@@ -10,7 +10,7 @@
 * `BatchingActivityProcessor`/`SimpleActivityProcessor` is disposable and it
   disposes the containing exporter.
 * `BroadcastActivityProcessor`is disposable and it disposes the processors.
-
+* Samplers now get the actual TraceId of the Activity to be created.
```
Comment (reyang) [inline src/OpenTelemetry/CHANGELOG.md:13]: I guess we need an empty line here.

Comment (cijothomas) [reply src/OpenTelemetry/CHANGELOG.md:13]: Your favourite tool caught it as well!

**Comment by alanwest** [comment]
Just want to clarify my understanding, next OTel release will coincide with preview 8. So, will preview 8 will not contain the removal of `AutoGenerateRootContextTraceId`? Or is this just getting the functionality/tests in place and then the plan is to change this prior to release?

**Comment by cijothomas** [comment]
> Just want to clarify my understanding, next OTel release will coincide with preview 8. So, will preview 8 will not contain the removal of `AutoGenerateRootContextTraceId`? Or is this just getting the functionality/tests in place and then the plan is to change this prior to release?

Preview8 won't remove `AutoGenerateRootContextTraceId`. It'll be RC1 or RC2.

File: src/OpenTelemetry/CHANGELOG.md (hunk 1 above)
Comment (CodeBlanch) [reply src/OpenTelemetry/CHANGELOG.md:13]: 🚨

**Comment by CodeBlanch** [comment]
@cijothomas @reyang Hey I just noticed something, this broke the `ParentOrElse` sampler.

```csharp
            var parentContext = samplingParameters.ParentContext;
            if (parentContext == default)
```

`parentContext == default` will now never be true, because a TraceId is always generated. Not sure how to detect "no parent" case now. Check SpanId = all 0s?

**Pull request merged**